File: apps/java/src/main/java/com/inthebeginning/simulator/ChemicalSystem.java
package com.inthebeginning.simulator;

import java.util.*;
import java.util.stream.Collectors;

import static com.inthebeginning.simulator.Constants.*;

/**
 * Manages molecular assembly and chemical reactions.
 * Forms water, methane, ammonia, amino acids, and nucleotides from available atoms.
 */
public class ChemicalSystem {

    private final AtomicSystem atomic;
    private final List<Molecule> molecules = new ArrayList<>();
    private int reactionsOccurred = 0;
    private int waterCount = 0;
    private int aminoAcidCount = 0;
    private int nucleotideCount = 0;
    private final Random rng;

    public ChemicalSystem(AtomicSystem atomic, Random rng) {
        this.atomic = atomic;
        this.rng = rng;
    }

    // --- Accessors ---

    public List<Molecule> getMolecules()    { return molecules; }
    public int getWaterCount()              { return waterCount; }
    public int getAminoAcidCount()          { return aminoAcidCount; }
    public int getNucleotideCount()         { return nucleotideCount; }
    public int getReactionsOccurred()       { return reactionsOccurred; }

    // --- Molecule formation ---

    /** Form water molecules: 2H + O -> H2O */
    public List<Molecule> formWater() {
        List<Molecule> waters = new ArrayList<>();
        List<Atom> hydrogens = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 1 && a.bonds().isEmpty())
                .collect(Collectors.toCollection(ArrayList::new));
        List<Atom> oxygens = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 8 && a.bonds().size() < 2)
                .collect(Collectors.toCollection(ArrayList::new));

        while (hydrogens.size() >= 2 && !oxygens.isEmpty()) {
            Atom h1 = hydrogens.remove(hydrogens.size() - 1);
            Atom h2 = hydrogens.remove(hydrogens.size() - 1);
            Atom o = oxygens.remove(oxygens.size() - 1);

            Molecule water = new Molecule(List.of(h1, h2, o), "water", o.position());
            waters.add(water);
            molecules.add(water);
            waterCount++;

            h1.bonds().add(o.atomId());
            h2.bonds().add(o.atomId());
            o.bonds().add(h1.atomId());
            o.bonds().add(h2.atomId());
        }
        return waters;
    }

    /** Form methane: C + 4H -> CH4 */
    public List<Molecule> formMethane() {
        List<Molecule> methanes = new ArrayList<>();
        List<Atom> carbons = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 6 && a.bonds().isEmpty())
                .collect(Collectors.toCollection(ArrayList::new));
        List<Atom> hydrogens = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 1 && a.bonds().isEmpty())
                .collect(Collectors.toCollection(ArrayList::new));

        while (!carbons.isEmpty() && hydrogens.size() >= 4) {
            Atom c = carbons.remove(carbons.size() - 1);
            List<Atom> hs = new ArrayList<>();
            for (int i = 0; i < 4; i++) hs.add(hydrogens.remove(hydrogens.size() - 1));

            List<Atom> allAtoms = new ArrayList<>();
            allAtoms.add(c);
            allAtoms.addAll(hs);
            Molecule methane = new Molecule(allAtoms, "methane", c.position());
            methanes.add(methane);
            molecules.add(methane);

            for (Atom h : hs) {
                c.bonds().add(h.atomId());
                h.bonds().add(c.atomId());
            }
        }
        return methanes;
    }

    /** Form ammonia: N + 3H -> NH3 */
    public List<Molecule> formAmmonia() {
        List<Molecule> ammonias = new ArrayList<>();
        List<Atom> nitrogens = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 7 && a.bonds().isEmpty())
                .collect(Collectors.toCollection(ArrayList::new));
        List<Atom> hydrogens = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 1 && a.bonds().isEmpty())
                .collect(Collectors.toCollection(ArrayList::new));

        while (!nitrogens.isEmpty() && hydrogens.size() >= 3) {
            Atom n = nitrogens.remove(nitrogens.size() - 1);
            List<Atom> hs = new ArrayList<>();
            for (int i = 0; i < 3; i++) hs.add(hydrogens.remove(hydrogens.size() - 1));

            List<Atom> allAtoms = new ArrayList<>();
            allAtoms.add(n);
            allAtoms.addAll(hs);
            Molecule ammonia = new Molecule(allAtoms, "ammonia", n.position());
            ammonias.add(ammonia);
            molecules.add(ammonia);

            for (Atom h : hs) {
                n.bonds().add(h.atomId());
                h.bonds().add(n.atomId());
            }
        }
        return ammonias;
    }

    /**
     * Form an amino acid from available atoms.
     * Simplified: requires 2C + 5H + 2O + 1N minimum (glycine-like).
     */
    public Molecule formAminoAcid(String aaType) {
        List<Atom> carbons = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 6 && a.bonds().isEmpty())
                .collect(Collectors.toCollection(ArrayList::new));
        List<Atom> hydrogens = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 1 && a.bonds().isEmpty())
                .collect(Collectors.toCollection(ArrayList::new));
        List<Atom> oxygens = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 8 && a.bonds().size() < 2)
                .collect(Collectors.toCollection(ArrayList::new));
        List<Atom> nitrogens = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 7 && a.bonds().isEmpty())
                .collect(Collectors.toCollection(ArrayList::new));

        if (carbons.size() < 2 || hydrogens.size() < 5 ||
            oxygens.size() < 2 || nitrogens.size() < 1) {
            return null;
        }

        List<Atom> atoms = new ArrayList<>();
        for (int i = 0; i < 2; i++) atoms.add(carbons.remove(carbons.size() - 1));
        for (int i = 0; i < 5; i++) atoms.add(hydrogens.remove(hydrogens.size() - 1));
        for (int i = 0; i < 2; i++) atoms.add(oxygens.remove(oxygens.size() - 1));
        atoms.add(nitrogens.remove(nitrogens.size() - 1));

        Molecule aa = new Molecule(atoms, aaType, atoms.get(0).position(),
                true, List.of("amino", "carboxyl"));
        molecules.add(aa);
        aminoAcidCount++;

        // Form internal bonds
        Atom first = atoms.get(0);
        for (int i = 1; i < atoms.size(); i++) {
            first.bonds().add(atoms.get(i).atomId());
            atoms.get(i).bonds().add(first.atomId());
        }

        return aa;
    }

    /**
     * Form a nucleotide (sugar + phosphate + base).
     * Simplified: requires C5 + H8 + O4 + N2.
     */
    public Molecule formNucleotide(String base) {
        List<Atom> carbons = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 6 && a.bonds().isEmpty())
                .collect(Collectors.toCollection(ArrayList::new));
        List<Atom> hydrogens = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 1 && a.bonds().isEmpty())
                .collect(Collectors.toCollection(ArrayList::new));
        List<Atom> oxygens = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 8 && a.bonds().size() < 2)
                .collect(Collectors.toCollection(ArrayList::new));
        List<Atom> nitrogens = atomic.getAtoms().stream()
                .filter(a -> a.atomicNumber() == 7 && a.bonds().isEmpty())
                .collect(Collectors.toCollection(ArrayList::new));

        if (carbons.size() < 5 || hydrogens.size() < 8 ||
            oxygens.size() < 4 || nitrogens.size() < 2) {
            return null;
        }

        List<Atom> atoms = new ArrayList<>();
        for (int i = 0; i < 5; i++) atoms.add(carbons.remove(carbons.size() - 1));
        for (int i = 0; i < 8; i++) atoms.add(hydrogens.remove(hydrogens.size() - 1));
        for (int i = 0; i < 4; i++) atoms.add(oxygens.remove(oxygens.size() - 1));
        for (int i = 0; i < 2; i++) atoms.add(nitrogens.remove(nitrogens.size() - 1));

        Molecule nuc = new Molecule(atoms, "nucleotide-" + base, atoms.get(0).position(),
                true, List.of("sugar", "phosphate", "base"));
        molecules.add(nuc);
        nucleotideCount++;

        Atom first = atoms.get(0);
        for (int i = 1; i < atoms.size(); i++) {
            first.bonds().add(atoms.get(i).atomId());
            atoms.get(i).bonds().add(first.atomId());
        }

        return nuc;
    }

    /**
     * Run catalyzed reactions to form complex molecules.
     * Activation energy is lower with catalyst.
     */
    public int catalyzedReaction(double temperature, boolean catalystPresent) {
        int formed = 0;
        double eaFactor = catalystPresent ? 0.3 : 1.0;
        double thermal = K_B * temperature;

        // Try to form amino acids
        if (thermal > 0 && atomic.getAtoms().size() > 10) {
            double aaProb = Math.exp(-5.0 * eaFactor / (thermal + 1e-20));
            if (rng.nextDouble() < aaProb) {
                String aaType = AMINO_ACIDS[rng.nextInt(AMINO_ACIDS.length)];
                if (formAminoAcid(aaType) != null) {
                    formed++;
                    reactionsOccurred++;
                }
            }
        }

        // Try to form nucleotides
        if (thermal > 0 && atomic.getAtoms().size() > 19) {
            double nucProb = Math.exp(-8.0 * eaFactor / (thermal + 1e-20));
            if (rng.nextDouble() < nucProb) {
                String[] bases = {"A", "T", "G", "C"};
                String base = bases[rng.nextInt(bases.length)];
                if (formNucleotide(base) != null) {
                    formed++;
                    reactionsOccurred++;
                }
            }
        }

        return formed;
    }

    /** Count molecules by type. */
    public Map<String, Integer> moleculeCensus() {
        Map<String, Integer> counts = new LinkedHashMap<>();
        for (Molecule m : molecules) {
            String key = (m.name() != null && !m.name().isEmpty()) ? m.name() : m.formula();
            counts.merge(key, 1, Integer::sum);
        }
        return counts;
    }

    public String toCompact() {
        Map<String, Integer> counts = moleculeCensus();
        StringBuilder sb = new StringBuilder();
        counts.entrySet().stream().sorted(Map.Entry.comparingByKey())
                .forEach(e -> {
                    if (sb.length() > 0) sb.append(",");
                    sb.append(e.getKey()).append(":").append(e.getValue());
                });
        return String.format("CS[mol=%d H2O=%d aa=%d nuc=%d rxn=%d %s]",
                molecules.size(), waterCount, aminoAcidCount,
                nucleotideCount, reactionsOccurred, sb);
    }
}

File: apps/kotlin/app/src/main/java/com/inthebeginning/simulator/ChemicalSystem.kt
package com.inthebeginning.simulator

import com.inthebeginning.simulator.Constants.BOND_ENERGY_COVALENT
import com.inthebeginning.simulator.Constants.BOND_ENERGY_HYDROGEN
import com.inthebeginning.simulator.Constants.BOND_ENERGY_VAN_DER_WAALS
import com.inthebeginning.simulator.Constants.K_B
import kotlin.math.exp
import kotlin.random.Random

/**
 * Type of chemical bond between atoms.
 */
enum class BondType(val displayName: String, val energy: Double) {
    COVALENT("covalent", BOND_ENERGY_COVALENT),
    IONIC("ionic", Constants.BOND_ENERGY_IONIC),
    HYDROGEN("hydrogen", BOND_ENERGY_HYDROGEN),
    VAN_DER_WAALS("van der Waals", BOND_ENERGY_VAN_DER_WAALS);
}

/**
 * A chemical bond between two atoms.
 */
data class Bond(
    val atom1Id: Int,
    val atom2Id: Int,
    val type: BondType,
    val strength: Double = type.energy
)

/**
 * A molecule composed of multiple bonded atoms.
 */
data class Molecule(
    val name: String,
    val formula: String,
    val atoms: List<Atom>,
    val bonds: List<Bond>,
    val moleculeId: Int = nextId()
) {
    companion object {
        private var idCounter = 0
        fun nextId(): Int = ++idCounter
        fun resetIds() { idCounter = 0 }
    }

    /** Total molecular mass. */
    val mass: Double get() = atoms.sumOf { it.massNumber.toDouble() }

    /** Number of atoms in the molecule. */
    val size: Int get() = atoms.size

    /** Total bond energy of the molecule. */
    val totalBondEnergy: Double get() = bonds.sumOf { it.strength }

    /** Check if this molecule contains a specific element. */
    fun containsElement(atomicNumber: Int): Boolean =
        atoms.any { it.atomicNumber == atomicNumber }

    /** Count occurrences of a specific element. */
    fun elementCount(atomicNumber: Int): Int =
        atoms.count { it.atomicNumber == atomicNumber }
}

/**
 * Predefined molecular templates for common molecules.
 */
sealed class MoleculeTemplate(val name: String, val formula: String) {
    /** Required atomic numbers and counts. */
    abstract val composition: Map<Int, Int>

    data object Water : MoleculeTemplate("Water", "H2O") {
        override val composition = mapOf(1 to 2, 8 to 1)
    }

    data object CarbonDioxide : MoleculeTemplate("Carbon Dioxide", "CO2") {
        override val composition = mapOf(6 to 1, 8 to 2)
    }

    data object Methane : MoleculeTemplate("Methane", "CH4") {
        override val composition = mapOf(6 to 1, 1 to 4)
    }

    data object Ammonia : MoleculeTemplate("Ammonia", "NH3") {
        override val composition = mapOf(7 to 1, 1 to 3)
    }

    data object HydrogenCyanide : MoleculeTemplate("Hydrogen Cyanide", "HCN") {
        override val composition = mapOf(1 to 1, 6 to 1, 7 to 1)
    }

    data object Formaldehyde : MoleculeTemplate("Formaldehyde", "CH2O") {
        override val composition = mapOf(6 to 1, 1 to 2, 8 to 1)
    }

    data object PhosphoricAcid : MoleculeTemplate("Phosphoric Acid", "H3PO4") {
        override val composition = mapOf(1 to 3, 15 to 1, 8 to 4)
    }

    data object Glycine : MoleculeTemplate("Glycine", "C2H5NO2") {
        override val composition = mapOf(6 to 2, 1 to 5, 7 to 1, 8 to 2)
    }

    data object Adenine : MoleculeTemplate("Adenine", "C5H5N5") {
        override val composition = mapOf(6 to 5, 1 to 5, 7 to 5)
    }

    data object Ribose : MoleculeTemplate("Ribose", "C5H10O5") {
        override val composition = mapOf(6 to 5, 1 to 10, 8 to 5)
    }
}

/**
 * Represents the type of an amino acid within the simulation.
 */
data class AminoAcidInfo(
    val code: String,
    val hydrophobic: Boolean,
    val charge: Int
)

/** Simplified amino acid properties. */
val AMINO_ACID_PROPERTIES: Map<String, AminoAcidInfo> = mapOf(
    "Ala" to AminoAcidInfo("A", hydrophobic = true, charge = 0),
    "Arg" to AminoAcidInfo("R", hydrophobic = false, charge = 1),
    "Asn" to AminoAcidInfo("N", hydrophobic = false, charge = 0),
    "Asp" to AminoAcidInfo("D", hydrophobic = false, charge = -1),
    "Cys" to AminoAcidInfo("C", hydrophobic = true, charge = 0),
    "Gln" to AminoAcidInfo("Q", hydrophobic = false, charge = 0),
    "Glu" to AminoAcidInfo("E", hydrophobic = false, charge = -1),
    "Gly" to AminoAcidInfo("G", hydrophobic = true, charge = 0),
    "His" to AminoAcidInfo("H", hydrophobic = false, charge = 1),
    "Ile" to AminoAcidInfo("I", hydrophobic = true, charge = 0),
    "Leu" to AminoAcidInfo("L", hydrophobic = true, charge = 0),
    "Lys" to AminoAcidInfo("K", hydrophobic = false, charge = 1),
    "Met" to AminoAcidInfo("M", hydrophobic = true, charge = 0),
    "Phe" to AminoAcidInfo("F", hydrophobic = true, charge = 0),
    "Pro" to AminoAcidInfo("P", hydrophobic = true, charge = 0),
    "Ser" to AminoAcidInfo("S", hydrophobic = false, charge = 0),
    "Thr" to AminoAcidInfo("T", hydrophobic = false, charge = 0),
    "Trp" to AminoAcidInfo("W", hydrophobic = true, charge = 0),
    "Tyr" to AminoAcidInfo("Y", hydrophobic = false, charge = 0),
    "Val" to AminoAcidInfo("V", hydrophobic = true, charge = 0),
)

/**
 * Represents a nucleotide (base + sugar + phosphate).
 */
data class Nucleotide(
    val base: String,
    val isRNA: Boolean = false
) {
    /** Watson-Crick complement. */
    val complement: String
        get() = when (base) {
            "A" -> if (isRNA) "U" else "T"
            "T" -> "A"
            "U" -> "A"
            "G" -> "C"
            "C" -> "G"
            else -> "?"
        }
}

/**
 * The chemical system manages molecular synthesis, reactions,
 * and the emergence of prebiotic chemistry.
 */
class ChemicalSystem(
    var temperature: Double = Constants.T_EARTH_SURFACE
) {
    val molecules: MutableList<Molecule> = mutableListOf()
    val aminoAcids: MutableList<String> = mutableListOf()
    val nucleotides: MutableList<Nucleotide> = mutableListOf()
    var reactionsPerformed: Int = 0

    private val rng = Random

    /**
     * Attempt to synthesize a molecule from available atoms in the atomic system.
     * Returns the molecule if successful, null otherwise.
     */
    fun synthesize(template: MoleculeTemplate, atomicSystem: AtomicSystem): Molecule? {
        val available = atomicSystem.atoms.groupBy { it.atomicNumber }
            .mapValues { it.value.toMutableList() }

        // Check if we have enough atoms
        for ((atomicNum, count) in template.composition) {
            val have = available[atomicNum]?.size ?: 0
            if (have < count) return null
        }

        // Check thermal feasibility
        val thermalEnergy = K_B * temperature
        val activationEnergy = BOND_ENERGY_COVALENT * 0.5
        val reactionProb = exp(-activationEnergy / thermalEnergy.coerceAtLeast(1e-10))
        if (rng.nextDouble() > reactionProb) return null

        // Consume atoms from the atomic system
        val usedAtoms = mutableListOf<Atom>()
        val bonds = mutableListOf<Bond>()
        for ((atomicNum, count) in template.composition) {
            val pool = available[atomicNum]!!
            repeat(count) {
                val atom = pool.removeAt(pool.lastIndex)
                atomicSystem.atoms.remove(atom)
                usedAtoms.add(atom)
            }
        }

        // Create bonds between consecutive atoms (simplified linear bonding)
        for (i in 0 until usedAtoms.size - 1) {
            val a1 = usedAtoms[i]
            val a2 = usedAtoms[i + 1]
            val bondType = when {
                kotlin.math.abs(a1.electronegativity - a2.electronegativity) > 1.7 -> BondType.IONIC
                else -> BondType.COVALENT
            }
            bonds.add(Bond(a1.atomId, a2.atomId, bondType))
        }

        val molecule = Molecule(
            name = template.name,
            formula = template.formula,
            atoms = usedAtoms,
            bonds = bonds
        )
        molecules.add(molecule)
        reactionsPerformed++
        return molecule
    }

    /**
     * Form water molecules from available H and O atoms.
     */
    fun formWater(atomicSystem: AtomicSystem): List<Molecule> {
        val formed = mutableListOf<Molecule>()
        var attempts = 0
        val maxAttempts = atomicSystem.atoms.size / 3 + 1
        while (attempts < maxAttempts) {
            val mol = synthesize(MoleculeTemplate.Water, atomicSystem) ?: break
            formed.add(mol)
            attempts++
        }
        return formed
    }

    /**
     * Attempt to form simple organic molecules (prebiotic chemistry).
     * Requires carbon, hydrogen, nitrogen, oxygen atoms in the atomic system.
     */
    fun prebioticSynthesis(atomicSystem: AtomicSystem): List<Molecule> {
        val formed = mutableListOf<Molecule>()

        // Try forming each type of prebiotic molecule
        val templates = listOf(
            MoleculeTemplate.Methane,
            MoleculeTemplate.Ammonia,
            MoleculeTemplate.HydrogenCyanide,
            MoleculeTemplate.Formaldehyde,
        )

        for (template in templates) {
            val mol = synthesize(template, atomicSystem)
            if (mol != null) formed.add(mol)
        }

        return formed
    }

    /**
     * Attempt amino acid synthesis from simpler molecules.
     * In reality this requires specific conditions (Miller-Urey experiment).
     */
    fun aminoAcidSynthesis(environment: Environment): List<String> {
        val formed = mutableListOf<String>()

        // Requires liquid water and energy source
        if (!environment.hasLiquidWater) return formed
        if (environment.radiation < 1.0 && !environment.hasLightning) return formed

        // UV radiation or lightning provides energy for synthesis
        val energyFactor = if (environment.hasLightning) 2.0 else 1.0
        val prob = 0.001 * energyFactor

        for (aminoAcid in Constants.AMINO_ACIDS) {
            if (rng.nextDouble() < prob) {
                aminoAcids.add(aminoAcid)
                formed.add(aminoAcid)
            }
        }

        return formed
    }

    /**
     * Attempt nucleotide synthesis from available molecules.
     * Requires phosphoric acid, ribose, and nucleobases.
     */
    fun nucleotideSynthesis(environment: Environment): List<Nucleotide> {
        val formed = mutableListOf<Nucleotide>()

        if (!environment.hasLiquidWater) return formed

        // Probability depends on temperature (warm ponds hypothesis)
        val tempFactor = if (temperature in 300.0..380.0) 2.0 else 0.5
        val prob = 0.0005 * tempFactor

        val bases = if (rng.nextDouble() < 0.5) {
            Constants.NUCLEOTIDE_BASES
        } else {
            Constants.RNA_BASES
        }

        for (base in bases) {
            if (rng.nextDouble() < prob) {
                val isRNA = base == "U" || (base != "T" && rng.nextDouble() < 0.7)
                val nucleotide = Nucleotide(base, isRNA)
                nucleotides.add(nucleotide)
                formed.add(nucleotide)
            }
        }

        return formed
    }

    /**
     * Polymerize nucleotides into short RNA chains (RNA World hypothesis).
     * Returns the length of the chain formed, or 0 if none.
     */
    fun polymerizeRNA(): Int {
        val rnaNucleotides = nucleotides.filter { it.isRNA }
        if (rnaNucleotides.size < 3) return 0

        // Probability of successful polymerization decreases with chain length
        val maxLen = rnaNucleotides.size.coerceAtMost(50)
        var chainLen = 0
        for (i in 0 until maxLen) {
            if (rng.nextDouble() < 0.3) {
                chainLen++
            } else {
                break
            }
        }

        // Remove used nucleotides
        repeat(chainLen.coerceAtMost(nucleotides.size)) {
            val idx = nucleotides.indexOfFirst { it.isRNA }
            if (idx >= 0) nucleotides.removeAt(idx)
        }

        return chainLen
    }

    /** Count molecules by name. */
    fun moleculeCounts(): Map<String, Int> {
        val counts = mutableMapOf<String, Int>()
        for (m in molecules) {
            counts[m.name] = (counts[m.name] ?: 0) + 1
        }
        return counts
    }

    /** Total number of distinct molecule types. */
    val moleculeDiversity: Int get() = molecules.map { it.name }.distinct().size
}
